src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpTypingStyleDiagnosticAnalyzerBase.State.cs

```diff
@@ -0,0 +1,237 @@
+
+            public TypingStyles StylePreferences { get; private set; }
+            public bool IsInIntrinsicTypeContext { get; private set; }
+            public bool IsTypingApparentInContext { get; private set; }
+            public bool IsInVariableDeclarationContext { get; }
+
+            public State(bool isVariableDeclarationContext)
+            {
+                this.IsInVariableDeclarationContext = isVariableDeclarationContext;
+                _styleToSeverityMap = new Dictionary<TypingStyles, DiagnosticSeverity>();
+            }
+
+            public static State Generate(SyntaxNode declaration, SemanticModel semanticModel, OptionSet optionSet, bool isVariableDeclarationContext, CancellationToken cancellationToken)
+            {
+                var state = new State(isVariableDeclarationContext);
+                state.Initialize(declaration, semanticModel, optionSet, cancellationToken);
+                return state;
+            }
+
+            public DiagnosticSeverity GetDiagnosticSeverityPreference()
+            {
+                if (IsInIntrinsicTypeContext)
+                {
+                    return _styleToSeverityMap[TypingStyles.VarForIntrinsic];
+                }
+                else if (IsTypingApparentInContext)
+                {
+                    return _styleToSeverityMap[TypingStyles.VarWhereApparent];
+                }
+                else
+                {
+                    return _styleToSeverityMap[TypingStyles.VarWherePossible];
+                }
+            }
+
+            public bool ShouldNotify() => 
+                GetDiagnosticSeverityPreference() != DiagnosticSeverity.Hidden;
+
+            private void Initialize(SyntaxNode declaration, SemanticModel semanticModel, OptionSet optionSet, CancellationToken cancellationToken)
+            {
+                this.StylePreferences = GetCurrentTypingStylePreferences(optionSet);
+
+                IsTypingApparentInContext =
+                    IsInVariableDeclarationContext
+                        ? IsTypeApparentInDeclaration((VariableDeclarationSyntax)declaration,
+                            semanticModel, StylePreferences, cancellationToken)
+                        : false;
+
+                IsInIntrinsicTypeContext = IsIntrinsicType(declaration);
+            }
+
+            /// <summary>
+            /// Returns true if type information could be gleaned by simply looking at the given statement.
+            /// This typically means that the type name occurs in either left hand or right hand side of an assignment.
+            /// </summary>
+            private bool IsTypeApparentInDeclaration(VariableDeclarationSyntax variableDeclaration, SemanticModel semanticModel, TypingStyles stylePreferences, CancellationToken cancellationToken)
+            {
+                var initializer = variableDeclaration.Variables.Single().Initializer;
```

Comment:
We cant use `var` on a declaration that has more than one variable, so they're filtered out before it reaches here. I can add an assert if you like.